Lib/email/_encoded_words.py

```diff
@@ -108,20 +108,25 @@ def decode_b(encoded):
     try:
         return base64.b64decode(padded_encoded, validate=True), defects
     except binascii.Error:
-        # Since we had correct padding, this must an invalid char error.
+        # Since we had correct padding, this is likely an invalid char error.
         defects = [errors.InvalidBase64CharactersDefect()]
         # The non-alphabet characters are ignored as far as padding
-        # goes, but we don't know how many there are.  So we'll just
-        # try various padding lengths until something works.
-        for i in 0, 1, 2, 3:
+        # goes, but we don't know how many there are.  So we'll try without
+        # adding padding to see if it works, and if not we'll add as much
+        # padding as could possibly be necessary (extra padding is ignored).
+        try:
+            return base64.b64decode(encoded, validate=False), defects
+        except binascii.Error:
+            defects.append(errors.InvalidBase64PaddingDefect())
             try:
-                return base64.b64decode(encoded+b'='*i, validate=False), defects
+                return base64.b64decode(encoded+b'==', validate=False), defects
             except binascii.Error:
-                if i==0:
-                    defects.append(errors.InvalidBase64PaddingDefect())
-        else:
-            # This should never happen.
-            raise AssertionError("unexpected binascii.Error")
+                # This only happens when the encoded string's length is 1 more
+                # than a multiple of 4, which is invalid.
+                defects = [errors.InvalidBase64LengthDefect()]
```

Comment:
This replacement of the defect list bothers me, it makes the code more fragile.

I think we should probably instead drop the 'pad_err' computation, since we're going to figure that out in the try/except, and then switch to computing the (value, defects) tuple and have a single return at the end of the method.  That way the PaddingDefect can be added in the else clause of that try/except, and we'll always be appending to the defect list.

Reply:
I've tried to address the code's fragility and opaqueness with some restructuring and improved comments.

Adding missing padding in the first attempt, which uses `validate=True`, is required to pass the validation in the case where all of the characters are valid but just padding is missing. This is different than the following attempts, where it is assumed that there are invalid characters, and decoded is attempted first with and then without adding padding to be able to tell whether padding was missing.